### PR TITLE
Combined rotting and temperature

### PR DESCRIPTION
**Pull request by Hirmuolio**
#### Summary
```SUMMARY: Infrastructure "Combines rotting with item temperature"```

#### Purpose of change
Currently item rotting is calculated from the enviroment temperature.
The main change is that item rotting is now calculated from item temperature.

#### Describe the solution
Main change: item rotting is calculated from item temperature instead of enviroment temperature.

Temperature and rotting are now calculated together. All items that rot also have temperature.

Temperature and rotting are now processed at same time every 10 minutes. They can go slightly out of sync if something sets one of them (such as heating) but that shouldn't be a big problem. It will just cause the rot to be calculated few minutes later.

Temperatures for items that were outside the reality bubble (for less than 2 days) are now calculated for the past with 1 hour steps. This fixes few edge cases that can happen when you return to an item after long time. 
For the time beyond 2 days outside the item is just assumed to be at the enviroment temperature.
This is copied from how rot used to be calculated for out of bubble items. (it probably works) 

Items do not rot while smoking. The old way for this was kind of nonsense since enviroment temperature decided how items rot while they are ebeing smoked (just at half speed). Trying to replace that with the new system seemed so annoying that I decided to just make items not rot while they are smoked.  
Alternative would be to make them rot at some constant rate (temperature) while smoking.
Edit: The same code is now used for "processing" instead (smoking and milling). Rot is not calculated in it still.

#### Additional context
There shouldn't really be many particularly noticeable changes. I just think it is more sensical approach and fixes some edge cases.

Items you carry rot quite a bit faster now. The old system used the map temperature so all carried items are +5 F warmer than before. Items on ground are usually at almost same temperature as enviroment so their rot is about same as before.

Out of map items get the heat from enviroment added to them (fires, convection). No more frozen items in fires when you return to them after long time. Only works for heat sources that are present when you return to the item. Assumes that the heat source was active for the whole duration since last seen.

### CODE REVIEW

File: src/cata_utility.cpp [hunk 1]
```diff
@@ -262,6 +262,11 @@ double temp_to_kelvin( double fahrenheit )
     return temp_to_celsius( fahrenheit ) + 273.15;
 }
 
+double kelvin_to_fahrenheit( double kelvin )
```
Comment (Night-Pryanik) [inline src/cata_utility.cpp]:
```suggestion
double kelvin_to_fahrenheit( double kelvin )
```

File: src/cata_utility.h [hunk 2]
```diff
@@ -250,12 +250,19 @@ double convert_volume( int volume, int *out_scale );
...
  * @return Temperature in degrees K.
  */
 double temp_to_kelvin( double fahrenheit );
 
+/**
+ * Convert a temperature from Kelvin to degrees Fahrenheit.
+ *
+ * @return Temperature in degrees C.
+ */
+double kelvin_to_fahrenheit( double kelvin );
```
Comment (Night-Pryanik) [inline src/cata_utility.h]:
```suggestion
double kelvin_to_fahrenheit( double kelvin );
```

File: src/item.cpp [hunk 3]
```diff
@@ -3808,7 +3808,71 @@ int item::spoilage_sort_order()
...
+
+    const int dsteps = dropoff - temperatures::freezing;
+    const int dstep = ( 35.91 * std::pow( 2.0, static_cast<float>( dropoff ) / 16.0 ) / dsteps );
+
+    if( temp < temperatures::freezing ) {
+        return 0;
+    } else if( temp > cutoff ) {
+        return cutoffrot;
+    } else if( temp < dropoff ) {
+        return ( ( temp - temperatures::freezing ) * dstep );
```
Comment (Night-Pryanik) [inline src/item.cpp]:
```suggestion
        return ( temp - temperatures::freezing ) * dstep;
```

File: src/item.cpp [hunk 4]
```diff
@@ -6932,27 +6974,99 @@ void item::apply_freezerburn()
...
+        insulation *= 1.5; // clothing provides inventory some level of insulation
+        temp += 5; // body heat increases inventory temperature
+    }
+
+    // process temperature and rot at most every 100_turns (10 min)
+    // If the item has had its temperature/rot set the two can be out of sync
+    // Rot happens slower than temperature changes so for most part last_temp_check dominates
+    // note we're also gated by item::processing_speed
+    time_duration smallest_interval = 100_turns;
+    if( now - last_temp_check >  smallest_interval ) {
```
Comment (Night-Pryanik) [inline src/item.cpp]:
```suggestion
    if( now - last_temp_check > smallest_interval ) {
```

**Comment by Hirmuolio** [comment]
The normal item processing seems to be working for both temperature and rotting.

Items you carry rot quite a bit faster now. Seems like the old system used the map temperature so all carried items are +5 F warmer than before. Items on ground are usually at almost same temperature as enviroment so their rot is about same as before.

Temperature unit tests are completely broken (the function they use doesn't even exist anymore).

Gorgon is failing for some unkown reason. Astyle isn't giving any edits to any files.

**Comment by Hirmuolio** [comment]
It is probably "complete". Will just need testing (and fixes if testing finds problems).

**Comment by Hirmuolio** [comment]
You know how there are two different item::process() functions. They have different number of parameters so they are separate.

Now they are about to have same number of parameters and things get slightly annoying... (it doesn't need temperature anymore)

There are also two separate has_rotten_away() too. I'm not going to touch that one now.

**Comment by Hirmuolio** [comment]
I believe this to be ready.

**Comment by ZhilkinSerg** [comment]
Got an exception on game start twice today (doesn't happen on each game start):

image

File: tests/temperature_test.cpp [hunk 5]
```diff
@@ -155,64 +169,75 @@ TEST_CASE( "Rate of temperature change" )
...
         // frozen
         CHECK( is_nearly( meat1.temperature, 27315000 ) );
+        CHECK( is_nearly( meat2.temperature, meat1.temperature ) );
         CHECK( meat1.item_tags.count( "FROZEN" ) );
+        CHECK( meat2.item_tags.count( "FROZEN" ) );
+
+        calendar::turn = to_turn<int>( calendar::turn + 45_minutes );
+        meat1.process_temperature_rot( 68, 1, tripoint( 0, 0, 0 ), nullptr );
+        meat2.process_temperature_rot( 68, 1, tripoint( 0, 0, 0 ), nullptr );
+
```
Comment (ZhilkinSerg) [inline tests/temperature_test.cpp]: Why following three lines are duplicates of previous three lines?

Comment (Hirmuolio) [reply tests/temperature_test.cpp]: First line moves time forwards.
The two next lines claculate the item temperature at the new time point.

The time delta between calculations needs to be less than 1 hour or the code will start calculating past temperature using weather and the result is completely different.

**Comment by int-ua** [comment]
Looks like cold labs were omitted again.
Related: #24208 #24205

Try spawning raw liver, freezing it in a minifreezer and placing it in z -1 ice lab for 4 days while being absent from area. I'm getting unfreezing and spoilage in 0.E release, do you?

**Pull request merged**